text/0517-io-os-reform.md

```diff
@@ -473,12 +649,229 @@ throughout IO, we can go on to explore the modules in detail.
+
+    // Whenever reading from `self`, push the bytes read to `out`
+    fn tee<W: Writer>(self, out: W) -> Tee<Self, W> { ... }
+}
+impl<T: Reader> ReaderExt for T {}
+
+trait WriterExt: Writer {
+    // Reify a borrowed writer as owned
+    fn by_ref<'a>(&'a mut self) -> ByRef<'a, Self> { ... }
+
+    // Whenever bytes are written to `self`, write them to `other` as well
+    fn carbon_copy<W: Writer>(self, other: W) -> CarbonCopy<Self, W> { ... }
+}
+impl<T: Writer> WriterExt for T {}
+
+// An adaptor converting an `Iterator<u8>` to a `Reader`.
+pub struct IterReader<T> { ... }
+```
+
+As with `std::iter`, these adapters are object unsafe an hence placed
+in an extension trait with a blanket `impl`.
+
+Note that the same `ByRef` type is used for both `Reader` and `Writer`
+-- and this RFC proposes to use it for `std::iter` as well. The
+insight is that there is no difference between the *type* used for
+by-ref adapters in any of these cases; what changes is just which
+trait defers through it. So, we propose to add the following to `core::borrow`:
+
+```rust
+pub struct ByRef<'a, Sized? T:'a> {
+    pub inner: &'a mut T
+}
+```
+
+which will allow `impl`s like the following in `core::io`:
+
+```rust
+impl<'a, W: Writer> Writer for ByRef<'a, W> {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> Result<uint, W::Err> { self.inner.write(buf) }
+
+    #[inline]
+    fn flush(&mut self) -> Result<(), W::Err> { self.inner.flush() }
+}
+```
+
+#### Free functions
+[Free functions]: #free-functions
+
+The current `std::io::util` module also includes a number of primitive
+readers and writers, as well as `copy`. These are updated as follows:
+
+```rust
+// A reader that yields no bytes
+fn empty() -> Empty;
+
+// A reader that yields `byte` repeatedly (generalizes today's ZeroReader)
+fn repeat(byte: u8) -> Repeat;
+
+// A writer that ignores the bytes written to it (/dev/null)
+fn sink() -> Sink;
+
+// Copies all data from a Reader to a Writer
+pub fn copy<E, R, W>(r: &mut R, w: &mut W) -> NonatomicResult<(), uint, E> where
+    R: Reader<Err = E>,
+    W: Writer<Err = E>
+```
+
+#### Seeking
+[Seeking]: #seeking
+
+The seeking infrastructure is largely the same as today's, except that
+`tell` is renamed to follow the RFC's design principles and the `seek`
+signature is refactored with more precise types:
+
+```rust
+pub trait Seek {
+    type Err;
+    fn position(&self) -> Result<u64, Err>;
+    fn seek(&mut self, pos: SeekPos) -> Result<(), Err>;
+}
+
+pub enum SeekPos {
+    FromStart(u64),
+    FromEnd(u64),
+    FromCur(i64),
+}
+```
+
+#### Buffering
+[Buffering]: #buffering
+
+The current `Buffer` trait will be renamed to `BufferedReader` for
+clarity (and to open the door to `BufferedWriter` at some later
+point):
+
+```rust
+pub trait BufferedReader: Reader {
+    fn fill_buf(&mut self) -> Result<&[u8], Self::Err>;
+    fn consume(&mut self, amt: uint);
+
+    // This should perhaps yield an iterator
```

Comment:
An optimized implementation can copy 16 bytes at once and check of the byte in parallel. How do you expect to get this performance with an iterator that looks at one byte at a time? To illustrate this let's compare the performance of memchr and iter().position():
```
test _memchr  ... bench:     59265 ns/iter (+/- 1536)
test position ... bench:   1373400 ns/iter (+/- 9820)
```